src/metabase/pulse/render/js_engine.clj

```diff
@@ -46,10 +46,16 @@
 (defn execute-fn-name
   "Executes `js-fn-name` in js context with args"
   ^Value [^Context context js-fn-name & args]
-  (let [fn-ref (.eval context "js" js-fn-name)
-        args   (into-array Object args)]
-    (assert (.canExecute fn-ref) (str "cannot execute " js-fn-name))
-    (.execute fn-ref args)))
+  ;; TODO: locking context is not ideal, but contexts are currently being shared with all threads and GraalVM doesn't
+  ;; support concurrent execution for js.
+  ;; There is a couple of idea we can try:
+  ;; - put a thread pool around context initialization
+  ;; - init a new context for each thread
```

Comment:
Using fresh contexts sounds like a nice approach if it's not too expensive.

Reply:
I'm afraid it's too expensive, as we could have multiple threads trying to create a new context at once and it becomes uncontrollable of how many contexts we create.

It'll be really bad because there is one context that takes 120MiB.

Reply:
I guess we could use a small pool of contexts then? Say allocate just-in-time up to N, and block if they're all busy.

Reply:
We had a discussion with Ngoc about this and then releasing them is a bit of work. And eating 400 mb of mem feels bad given that lot of metabases are hosted on 2gb of memory.

Also: just doing it sync doesn't really degrade any experience as everything was serial before Ngoc's change. So I think just fixing the exact problem makes sense for now.

Reply:
Yeah agree this is good to go, just wanted to poke on whether we're happy with this TODO message.

It would be nice to have data on whether this is ever a bottleneck. Perhaps a first (future) step should be to track wait time analytics.

Reply:
I did something similar at first in https://github.com/metabase/metabase/commit/2dad954422b008c3a276b3fc4ec44a3b2dbb41e5, but I wanted to have something safe for us to start with. Then I'll discuss with the team about a better, more complex idea as part of solving https://github.com/metabase/metabase/issues/42496